### DISCUSSION TITLE
Could there be a  `CatalogItem` in HydroMT?

### DISCUSSION THREAD

**Discussion by Jaapel** [Deltares/hydromt#608] [RFC]
When attempting to implementing some nodata strategy for: https://github.com/Deltares/hydromt/issues/584 - valid for all different `DataAdapters`, I see that there are many calls that are very similar calls and argument passing going on.
Like `DataCatalog.get_<type>data()`, which all have similar arguments. Could we not create a `CatalogItem` in HydroMT, which would then be subclassed with `RasterItem`? We could then create (abstract) functionality that would work for all items, like `isempty`, `handle_empty`, and simplify getting the data.

**Comment by DirkEilander** [comment]
Could that also be handled by the DataAdapter and subclasses? An initialized DataAdapter instance is basically a CatalogItem.

**Reply by Jaapel** [reply]
True, but I feel like there is some duplication in the `get_rasterdataset` etc. functions.

**Reply by DirkEilander** [reply]
I agree there is room to abstract / generalize functionality! but I wonder if we need a separate class for it. Do you think this could also be handled within the current `DataAdapter` implementation?

**Reply by DirkEilander** [reply]
To clarify: I'm not against this idea and i'd be happy to see a more worked out proposal to understand it's potential.

**Reply by Jaapel** [reply]
I have not yet given it more thought. I think that will change once talk starts about HydroMT 1.0 ;). As for the issue, duplicated some logic across different methods.

**Comment by aslowwriter** [comment]
I think this is a good one to consider for our 1.0 refactor/reorganisation

**Comment by aslowwriter** [comment]
Having thought about this a bit more, I'm wondering if this is not just the generic `data_adapter` class that we already have. Couldn't we integrate more of the functionality there?
